sycl/source/device_selector.cpp

```diff
@@ -157,5 +161,152 @@ int host_selector::operator()(const device &dev) const {
+      Result.DeviceType = PI_DEVICE_TYPE_GPU;
+      Result.hasDeviceType = true;
+    } else if (Token == "accelerator" && !Result.hasDeviceType) {
+      Result.DeviceType = PI_DEVICE_TYPE_ACC;
+      Result.hasDeviceType = true;
+    } else if (Token == "opencl" && !Result.hasBackend) {
+      Result.Backend = backend::opencl;
+      Result.hasBackend = true;
+    } else if (Token == "level-zero" && !Result.hasBackend) {
+      Result.Backend = backend::level_zero;
+      Result.hasBackend = true;
+    } else if (Token == "cuda" && !Result.hasBackend) {
+      Result.Backend = backend::cuda;
+      Result.hasBackend = true;
+    } else if (Token == "host") {
+      if (!Result.hasBackend) {
+        Result.Backend = backend::host;
+        Result.hasBackend = true;
+      } else if (!Result.hasDeviceType && Result.Backend != backend::host) {
+        // We already set everything earlier or it's an error.
+        throw runtime_error("Cannot specify host device with non-host backend.",
+                            PI_INVALID_VALUE);
+      }
+    } else if (std::regex_match(Token, IntegerExpr) && !Result.hasDeviceNum) {
+      Result.DeviceNum = std::stoi(Token);
+      Result.hasDeviceNum = true;
+    } else {
+      throw runtime_error(Error, PI_INVALID_VALUE);
+    }
+  }
+
+  return Result;
+}
+} // namespace detail
+
+filter_selector::filter_selector(std::string filter)
+    : mFilters(), mRanker(), mNumDevicesSeen(0), mMatchFound(false) {
+  std::vector<std::string> Filters = detail::tokenize(filter, ",");
+  mNumTotalDevices = device::get_devices().size();
+
+  for (const std::string &Filter : Filters) {
+    detail::filter F = detail::create_filter(Filter);
+    mFilters.push_back(std::make_shared<detail::filter>(F));
+  }
+}
+
+int filter_selector::operator()(const device &dev) const {
+  int Score = REJECT_DEVICE_SCORE;
+
+  for (auto &Filter : mFilters) {
+    bool BackendOK = true;
+    bool DeviceTypeOK = true;
+    bool DeviceNumOK = true;
+
+    // handle host device specially
+    if (Filter->hasBackend) {
+      backend BE;
+      if (dev.is_host()) {
+        BE = backend::host;
+      } else {
+        BE = sycl::detail::getSyclObjImpl(dev)->getPlugin().getBackend();
+      }
+      BackendOK = (BE == Filter->Backend);
+    }
+    if (Filter->hasDeviceType) {
+      RT::PiDeviceType DT =
+          sycl::detail::getSyclObjImpl(dev)->get_device_type();
+      DeviceTypeOK = (DT == Filter->DeviceType);
+    }
+    if (Filter->hasDeviceNum) {
+      // Only check device number if we're good on the previous matches
+      if (BackendOK && DeviceTypeOK) {
+        // Do we match?
+        DeviceNumOK = (Filter->MatchesSeen == Filter->DeviceNum);
+        // Safe to increment matches even if we find it
+        Filter->MatchesSeen++;
+      }
+    }
+    if (BackendOK && DeviceTypeOK && DeviceNumOK) {
+      Score = mRanker(dev);
+      mMatchFound = true;
+      break;
+    }
+  }
+
+  mNumDevicesSeen++;
```

Comment:
I guess this value can be a problem if a user uses the same `filter_selector` variable several times.

Reply:
Just to clarify:
I think there can be a race if a user creates `sycl::device`s in multiple threads using the same value of `filter_selector`.
Also the same problem can happen if user copies value of `filter_selector` and uses original and newly got variable in multiple threads because they share the same `filter` structure.
I think this is blocking issue.

Reply:
The way to fix this is to move to sycl 2020 style device selectors.  Then I can force that an instance of a filter_selector can only be used once.